cmd/jag/commands/watch.go

```diff
@@ -125,29 +126,37 @@ func (w *watcher) CountPaths() int {
 }
 
 func (w *watcher) Watch(paths ...string) (err error) {
+	w.Mutex.Lock()
+	defer w.Mutex.Unlock()
+
 	for i, p := range paths {
 		if paths[i], err = filepath.EvalSymlinks(p); err != nil {
 			return err
 		}
 	}
 
+	candidateDirs := map[string]struct{}{}
 	candidates := map[string]struct{}{}
 	for _, p := range paths {
-		if _, ok := w.paths[p]; !ok {
-			w.Mutex.Lock()
-			w.watcher.Add(p)
-			w.paths[p] = struct{}{}
-			w.Mutex.Unlock()
+		dir := filepath.Dir(p)
+		w.paths[p] = struct{}{}
+		if _, ok := w.dirs[dir]; !ok {
+			w.watcher.Add(dir)
 		}
+		candidateDirs[dir] = struct{}{}
 		candidates[p] = struct{}{}
 	}
 
+	// Remove the files/watchers we don't need anymore.
 	for p := range w.paths {
 		if _, ok := candidates[p]; !ok {
-			w.Mutex.Lock()
-			w.watcher.Remove(p)
 			delete(w.paths, p)
```

Comment:
Check that deleting while iterating is okay and add a comment.

Reply:
It is: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop